app/sample_manifest_excel/sample_manifest_excel/upload/processor/tube_rack.rb

```diff
@@ -0,0 +1,201 @@
+
+            @tube_rack_information_processed = true
+          end
+          update_samples_and_aliquots(tag_group)
+          update_sample_manifest
+        end
+
+        def retrieve_tube_rack_barcodes_from_manifest
+          rack_barcodes_list = @upload.data.description_info.select { |key| key.start_with?('Rack barcode (') }.values
+          return nil if rack_barcodes_list.any?(nil)
+          return rack_barcodes_list
+        end
+
+        # if a tube rack record already exists for any of the rack barcodes in the manifest,
+        # it has been processed before and should not be re-processed
+        def should_process_tube_rack_information?
+          @tube_rack_barcodes_from_manifest.each do |barcode|
+            existing_barcode_record = Barcode.includes(:asset).find_by(barcode: barcode)
+            return false if !existing_barcode_record.nil? && !existing_barcode_record.asset.nil?
+          end unless @tube_rack_barcodes_from_manifest.nil?
+          true
+        end
+
+        def retrieve_scan_results
+          @rack_barcode_to_scan_results = {}
+          @tube_barcode_to_rack_barcode = {}
+          @tube_rack_barcodes_from_manifest.each do |tube_rack_barcode|
+            results = if Rails.configuration.do_tube_rack_scan_callout
+                        retrieve_tube_rack_scan_from_microservice(tube_rack_barcode)
+                      else
+                        mock_scan_result
+                      end
+            return false if results.nil?
+
+            @rack_barcode_to_scan_results[tube_rack_barcode] = results
+            results.keys.each { |tube_barcode| @tube_barcode_to_rack_barcode[tube_barcode] = tube_rack_barcode }
+          end
+
+          true
+        end
+
+        def retrieve_tube_rack_scan_from_microservice(tube_rack_barcode)
+          # tube_rack_barcode = 'test_valid_file'
+          host_name = Rails.configuration.tube_rack_scans_microservice_endpoint
+          path = '/tube_rack/' + tube_rack_barcode
+          port = Rails.configuration.tube_rack_scans_microservice_port
+          response = Net::HTTP.get_response(host_name, path, port)
+
+          begin
+            scan_results = JSON.parse(response.body)
+          rescue JSON::JSONError => e
+            error_message = "Response when trying to retrieve scan (tube rack with barcode #{tube_rack_barcode}) was not valid JSON so could not be understood. Error message: #{e.message}"
+            upload.errors.add(:base, error_message)
+            return nil
+          end
+
+          unless response.code == '200'
+            error_message = "Scan could not be retrieved for tube rack with barcode #{tube_rack_barcode}. Service responded with status code #{response.code} "
+            error_message += " and the following message: #{scan_results['error']}"
+            upload.errors.add(:base, error_message)
+            return nil
+          end
+
+          scan_results['layout'] || nil
+        end
```

Comment:
I would move all the functionality about talking to the microservice to its own client file (in the lib folder).